src/server/pkg/storage/fileset/reader.go

```diff
@@ -1,43 +1,56 @@
 package fileset
 
 import (
-	"bytes"
 	"context"
 	"io"
 
 	"github.com/gogo/protobuf/proto"
-	"github.com/pachyderm/pachyderm/src/client/pkg/errors"
 	"github.com/pachyderm/pachyderm/src/server/pkg/storage/chunk"
 	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/index"
-	"github.com/pachyderm/pachyderm/src/server/pkg/tar"
 )
 
 // Reader is an abstraction for reading a fileset.
 type Reader struct {
-	chunks       *chunk.Storage
-	indexFactory func(ctx context.Context) (*index.Reader, error)
+	store              Store
+	chunks             *chunk.Storage
+	path               string
+	indexOpts          []index.Option
+	additive, deletive *index.Reader
 }
 
 func newReader(store Store, chunks *chunk.Storage, p string, opts ...index.Option) *Reader {
-	return &Reader{
-		chunks: chunks,
-		indexFactory: func(ctx context.Context) (*index.Reader, error) {
-			topIdx, err := store.GetIndex(ctx, p)
-			if err != nil {
-				return nil, err
-			}
-			return index.NewReader(chunks, topIdx, opts...), nil
-		},
+	r := &Reader{
+		store:     store,
+		chunks:    chunks,
+		path:      p,
+		indexOpts: opts,
 	}
+	return r
 }
 
-// Iterate iterates over the files in the fileset.
-func (r *Reader) Iterate(ctx context.Context, cb func(File) error) error {
-	ir, err := r.indexFactory(ctx)
-	if err != nil {
+func (r *Reader) setup(ctx context.Context) error {
+	if r.additive == nil {
+		md, err := r.store.Get(ctx, r.path)
+		if err != nil {
+			return err
+		}
+		r.additive = index.NewReader(r.chunks, md.Additive, r.indexOpts...)
+		r.deletive = index.NewReader(r.chunks, md.Deletive, r.indexOpts...)
+	}
+	return nil
+}
+
+// Iterate iterates over the files in the file set.
+func (r *Reader) Iterate(ctx context.Context, cb func(File) error, deletive ...bool) error {
+	if err := r.setup(ctx); err != nil {
 		return err
 	}
-	return ir.Iterate(ctx, func(idx *index.Index) error {
+	if len(deletive) > 0 && deletive[0] {
```

Comment:
Thinking about this more, we should probably have a second method `IterateDeletes`.

Reply:
RE: PPS `Iterate` vs `ForEach` this might be a good chance to make that change.

Reply:
If we are introducing `NewIterator` I definitely think we should change these methods to `ForEach`.  Iterate would seems like it should return an object similar to `NewIterator`

Reply:
Refer to the first thread for `IterateDeletes`, and I think we should separate the global name change from this set of changes.

Reply:
I'm fine with adding this as a TODO.